File: 1014-k-closest-points-to-origin/1014-k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> heap = new PriorityQueue<>(Comparator.comparingInt(a -> (a[0] * a[0] + a[1] * a[1])));

        for (int[] point: points) {
            heap.offer(point);
        }

        int[][] result = new int[k][2];
        while (k > 0) {
            result[--k] = heap.poll();
        }

        return result;

    }
}

File: 1014-k-closest-points-to-origin/1014-k-closest-points-to-origin.py
def calculate(point: List[int]):
    return point[0] * point[0] + point[1] * point[1]


class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        dists = []
        heap = []

        for point in points:
            dist = calculate(point)
            heapq.heappush(heap, dist)
            dists.append(dist)

        kth_dist = [heapq.heappop(heap) for _ in range(k)][-1]

        return [points[idx] for idx, dist in enumerate(dists) if dist <= kth_dist]